### PR TITLE
feat(cloudfront): add support for Origin Groups

### LINKED ISSUE #9109: [CloudFront] Distribution - Origin Groups
Create support for Origin Groups (fallback origin configurations) for the new `Distribution` construct.

### Proposed Solution

See aws/aws-cdk-rfcs#171.

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by skinny85**
Fixes #9109

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by skinny85** [comment]
Forgot to update the ReadMe... nice work automatic checker!

**Comment by skinny85** [comment]
> Forgot to update the ReadMe... nice work automatic checker!

Done!

File: packages/@aws-cdk/aws-cloudfront-origins/README.md [hunk 1]
```diff
@@ -76,3 +76,20 @@ new cloudfront.Distribution(this, 'myDist', {
 ```
 
 See the documentation of `@aws-cdk/aws-cloudfront` for more information.
+
+## Failover Origins
```
Comment (njlynch) [inline packages/@aws-cdk/aws-cloudfront-origins/README.md:80]: I feel this heading should at least parenthetically reference "Origin Groups", as that's what the feature is (inconsistently) called in CloudFront.
```suggestion
## Failover Origins (Origin Groups)
```

File: packages/@aws-cdk/aws-cloudfront-origins/README.md [hunk 2]
```diff
@@ -76,3 +76,20 @@ new cloudfront.Distribution(this, 'myDist', {
...
+
```
Comment (njlynch) [inline packages/@aws-cdk/aws-cloudfront-origins/README.md:81]: I'd start this section with a brief (stolen) documentation section explaining what failover origins are. 

Something like this paragraph (perhaps slightly tweaked):
```suggestion
You can set up CloudFront with origin failover for scenarios that require high availability. To get started, you create an origin group with two origins: a primary and a secondary. If the primary origin is unavailable, or returns specific HTTP response status codes that indicate a failure, CloudFront automatically switches to the secondary origin.

```
*Source: https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/high_availability_origin_failover.html*

File: packages/@aws-cdk/aws-cloudfront-origins/README.md [hunk 3]
```diff
@@ -76,3 +76,20 @@ new cloudfront.Distribution(this, 'myDist', {
...
+If you want to provide a failover Origin,
```
Comment (njlynch) [inline packages/@aws-cdk/aws-cloudfront-origins/README.md:82]: *Nit:* Why the 40-character wrap? :)
```suggestion
If you want to provide a failover Origin
```

File: packages/@aws-cdk/aws-cloudfront/lib/distribution.ts [hunk 4]
```diff
@@ -197,10 +199,38 @@ export class Distribution extends Resource implements IDistribution {
...
+        }
+        const failoverOriginId = this.addOrigin(originBindConfig.failoverConfig.failoverOrigin, true);
+        this.addOriginGroup(originBindConfig.failoverConfig.statusCodes, originId, failoverOriginId);
+      }
       return originId;
     }
   }
 
+  private addOriginGroup(statusCodes: number[] | undefined, originId: string, failoverOriginId: string): void {
+    statusCodes = statusCodes ?? [500, 502, 503, 504];
```
Comment (njlynch) [inline packages/@aws-cdk/aws-cloudfront/lib/distribution.ts:214]: *Nit:* It feels a little weird to do this here, as opposed to in the OriginGroup, but I suppose this is more "future-proof" in terms of catching it at the latest point.

Question - Is it valid for the list of status codes to be empty (i.e., in the case where you only want to failover when the connection fails)? According to https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/high_availability_origin_failover.html, 

```
When any of the following occur:
* The primary origin returns an HTTP status code that you’ve configured for failover
* CloudFront fails to connect to the primary origin
* The response from the primary origin takes too long (times out)
Then CloudFront routes the request to the secondary origin in the origin group.
```

So I _think_ that means an empty list of status codes is valid and represents only failing over on connection issues? If so, then this should be ok; if not, we'd want to guard against that.

File: packages/@aws-cdk/aws-cloudfront/lib/distribution.ts [hunk 5]
```diff
@@ -187,7 +189,7 @@ export class Distribution extends Resource implements IDistribution {
     this.additionalBehaviors.push(new CacheBehavior(originId, { pathPattern, ...behaviorOptions }));
   }
 
-  private addOrigin(origin: IOrigin): string {
+  private addOrigin(origin: IOrigin, addingAFailoverOrigin: boolean = false): string {
```
Comment (njlynch) [inline packages/@aws-cdk/aws-cloudfront/lib/distribution.ts:192]: I like this slightly better, but don't feel super-strongly about it.

```suggestion
  private addOrigin(origin: IOrigin, isFailoverOrigin: boolean = false): string {
```

File: packages/@aws-cdk/aws-cloudfront-origins/README.md (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-cloudfront-origins/README.md:80]: Done!

File: packages/@aws-cdk/aws-cloudfront-origins/README.md (hunk 3 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-cloudfront-origins/README.md:82]: This is my personal style: in m opinion, docs should end lines on sentence boundaries. This makes future changes to these docs have much better diff than if you just follow a set line length like 80 or 120.

File: packages/@aws-cdk/aws-cloudfront-origins/README.md (hunk 2 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-cloudfront-origins/README.md:81]: Done!

File: packages/@aws-cdk/aws-cloudfront/lib/distribution.ts (hunk 5 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-cloudfront/lib/distribution.ts:192]: Sure. Changed.

File: packages/@aws-cdk/aws-cloudfront/lib/distribution.ts (hunk 4 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-cloudfront/lib/distribution.ts:214]: I tried deploying a stack with an empty `fallbackStatusCodes`, unfortunately it does not work:

```
1 validation error detected: 
Value '[]' at 'distributionConfigWithTags.distributionConfig.originGroups.items.1.member.failoverCriteria.statusCodes.items' 
failed to satisfy constraint: 
Member must have length greater than or equal to 1
(Service: AmazonCloudFront; Status Code: 400; Error Code: MalformedXML; Request ID: ee8399fa-968c-4122-8fc9-ef60cb2cf49c)
```

File: packages/@aws-cdk/aws-cloudfront-origins/README.md [hunk 6]
```diff
@@ -76,3 +76,24 @@ new cloudfront.Distribution(this, 'myDist', {
...
+You achieve that behavior in the CDK using the `OriginGroup` class:
+
+```ts
+new cloudfront.Distribution(this, 'myDist', {
+  defaultBehavior: {
+    origin: new origins.OriginGroup({
+      primaryOrigin: new origins.S3Origin(myBucket),
+      fallbackOrigin: new origins.HttpOrigin('www.example.com'),
+      // optional, defaults to: 500, 502, 503 and 504
+      fallbackStatusCodes: [501],
```
Comment (njlynch) [inline packages/@aws-cdk/aws-cloudfront-origins/README.md:95]: I don't believe 501 is supported.

> You can choose any combination of the following status codes: 500, 502, 503, 504, 404, or 403.

(Source: https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/high_availability_origin_failover.html)
```suggestion
      fallbackStatusCodes: [500],
```

Comment (skinny85) [reply packages/@aws-cdk/aws-cloudfront-origins/README.md:95]: Oh wow, did not know that, thanks!

**Pull request merged**